File: lovr/gt/lib/GrandTour.lua
local linalg = require('lib/linalg')

local GrandTour = {}

function GrandTour:new(ndim, stepsize)
	local ret = {}
	setmetatable(ret, self)
    self.__index = self

    ret.t = 0
	ret.ndim = ndim
	ret.stepsize = stepsize
	ret.thetas = {}
	ret.matrix = linalg.eye(ndim, ndim)
	for i=1,ndim do
		ret.thetas[i] = {}
		for j=1,ndim do
			ret.thetas[i][j] = math.random()-0.5
		end
	end
	return ret
end

function _rotate(matrix, i,j, theta)
	local tmp = {}
	tmp[1] = {}
	tmp[2] = {}
	for k=1,#matrix do
		tmp[1][k] = matrix[k][i]*math.cos(theta) + matrix[k][j]*math.sin(theta)
		tmp[2][k] = -matrix[k][i]*math.sin(theta) + matrix[k][j]*math.cos(theta)
	end

	for k=1,#matrix do
		matrix[k][i] = tmp[1][k]
		matrix[k][j] = tmp[2][k]
	end
	return matrix
end	

function GrandTour:tick(dt)
	self.t = self.t + dt
	for i=1,self.ndim do
		for j=i+1,self.ndim do
			self.matrix = _rotate(self.matrix, i,j, dt*self.thetas[i][j]*self.stepsize)
		end
	end
	return self.matrix
end

function GrandTour:project(data)
	self.points = linalg.dot(data, self.matrix)
	return self.points
end

--test
-- local m = linalg.eye(3,3)
-- for _=1,100 do
-- 	m = _rotate(m, 1,2,30)
-- 	for i=1,3 do
-- 		for j=1,3 do
-- 			io.write(tostring(m[i][j]))
-- 			io.write('\t')
-- 		end
-- 		io.write('\n')
-- 	end
-- 	print()
-- end

-- local gt = Grandtour:new(5, 0.5)
-- for i=1,100 do
-- 	gt:tick(1)
-- 	print(gt.matrix[1][1])
-- end

return GrandTour

File: npm/src/js/GrandTour.js
import math from 'mathjs';
// import numeric from 'numeric';

export default class GrandTour{
	constructor({ndim, stepsize=0.00007}){
		this.ndim = ndim;
 		this.stepsize0 = stepsize;
 		this.stepsize = stepsize;
	}

	get ndim(){
		return this._ndim;
	}
	set ndim(newNdim){
    	if(newNdim !== this.ndim){
	      	this._ndim = newNdim;
	      	this.N = this._ndim * this._ndim;
	      	this.initThetas(this.N);
    	}
	}


	initThetas(N){
		this._thetas = new Array(N);
		for(var i=0; i<N; i++){
			this._thetas[i] = (Math.random()-0.5) * Math.PI;
		}
		this._matrix = math.identity(this.ndim)._data;
	}


	get thetas(){
		return this._thetas;
	}


	rotateMatrix(matrix, i,j, theta){
		var sin = Math.sin(theta);
		var cos = Math.cos(theta);
		// var res = matrix.map(d=>d.slice());
		var column_i = matrix.map(d=>d[i]);
		var column_j = matrix.map(d=>d[j]);

		for(let rowIndex=0; rowIndex<matrix.length; rowIndex++){
			matrix[rowIndex][i] = column_i[rowIndex]*cos + column_j[rowIndex]*(-sin);
			matrix[rowIndex][j] = column_i[rowIndex]*sin + column_j[rowIndex]*cos;
		}
		return matrix;
	};


	tick(dt=0){
		let angles = this.thetas.map( theta => theta * dt * this.stepsize );
		var k = -1;
		for(var i=0; i<this.ndim; i++){
			for(var j=0; j<this.ndim; j++){
				if(i!==j && (true || i<=3 || j<=3) ){
					k++;
					this._matrix = this.rotateMatrix(this._matrix, i,j, angles[k]);
				}
			}
		}
		return this._matrix;
	}


	get matrix(){
		return this._matrix;
	}

	set matrix(newMatrix){
		this._matrix = newMatrix;
	}


	project(data, dt=0){
		this.tick(dt);
		var matrix = this._matrix;
		matrix = matrix.map((row)=>row.slice(0,3));
		var res = math.multiply(data, matrix);
		return res;
	};

}

